File: frontend/src/api/cashflow.ts
import { api } from "./client";

// ── Types ──────────────────────────────────────────────────────────────────

export interface WalletAccount {
  id: number;
  name: string;
  account_type: "bank" | "cash" | "ewallet" | "credit_card" | "other";
  owner_type: "company" | "personal" | "mixed";
  bank_name?: string;
  account_number?: string;
  account_holder?: string;
  currency: string;
  opening_balance: number;
  current_balance: number;
  is_active: boolean;
  notes?: string;
  created_at: string;
}

export interface Holder {
  id: number;
  name: string;
  holder_type: "company" | "personal" | "tax" | "salary" | "project" | "reserve" | "stock" | "other";
  owner_type: "company" | "personal" | "mixed";
  wallet_account_id?: number;
  purpose?: string;
  opening_balance: number;
  current_balance: number;
  responsible_user_id?: number;
  is_active: boolean;
  notes?: string;
  created_at: string;
}

export interface CashflowCategory {
  id: number;
  type: "income" | "expense" | "payable" | "receivable";
  name: string;
  parent_id?: number;
  color?: string;
  icon?: string;
  sort_order: number;
  is_active: boolean;
}

export interface IncomeEntry {
  id: string;
  income_date: string;
  document_no?: string;
  income_type?: string;
  category_id?: number;
  customer_name?: string;
  description: string;
  amount: number;
  vat_amount: number;
  withholding_tax: number;
  net_amount: number;
  payment_channel?: string;
  wallet_account_id?: number;
  holder_id?: number;
  status: "pending" | "completed" | "cancelled";
  received_date?: string;
  owner_type: string;
  notes?: string;
  created_at: string;
}

export interface ExpenseEntry {
  id: string;
  expense_date: string;
  document_no?: string;
  expense_type?: string;
  category_id?: number;
  vendor_name?: string;
  description: string;
  amount: number;
  vat_amount: number;
  withholding_tax: number;
  net_amount: number;
  payment_channel?: string;
  wallet_account_id?: number;
  holder_id?: number;
  is_company_expense: boolean;
  status: "pending" | "completed" | "cancelled";
  paid_date?: string;
  owner_type: string;
  notes?: string;
  created_at: string;
}

export interface Payable {
  id: string;
  creditor_name: string;
  creditor_type?: string;
  description?: string;
  issue_date: string;
  due_date?: string;
  total_amount: number;
  paid_amount: number;
  remaining_amount: number;
  expected_account_id?: number;
  expected_holder_id?: number;
  status: "unpaid" | "partial" | "paid" | "overdue" | "cancelled";
  reference_doc?: string;
  notes?: string;
  created_at: string;
}

export interface Receivable {
  id: string;
  debtor_name: string;
  debtor_type?: string;
  description?: string;
  issue_date: string;
  due_date?: string;
  total_amount: number;
  received_amount: number;
  remaining_amount: number;
  expected_account_id?: number;
  expected_holder_id?: number;
  status: "unreceived" | "partial" | "received" | "overdue" | "cancelled";
  reference_doc?: string;
  notes?: string;
  created_at: string;
}

export interface Transfer {
  id: string;
  transfer_date: string;
  transfer_type: string;
  from_account_id?: number;
  from_holder_id?: number;
  to_account_id?: number;
  to_holder_id?: number;
  amount: number;
  fee: number;
  reason?: string;
  status: string;
  notes?: string;
  created_at: string;
}

export interface BankReconciliationAccount {
  id: number;
  name: string;
  bank_name?: string;
  account_number?: string;
  currency: string;
  current_balance: number;
  period_start: string;
  period_end: string;
  total_count: number;
  reconciled_count: number;
  progress: number;
  last_import_at?: string;
}

export interface ReconciliationCashTransaction {
  id: number;
  transaction_date: string;
  direction: "in" | "out" | "transfer_in" | "transfer_out";
  amount: number;
  description?: string;
  reference_type: string;
  reference_id?: string;
  document_no?: string;
  document_count: number;
}

export interface BankStatementLine {
  id: number;
  transaction_date: string;
  transaction_time?: string;
  description: string;
  reference?: string;
  channel?: string;
  amount: number;
  status: "suggested" | "waiting" | "completed";
  suggested_score?: number;
  cash_transaction?: ReconciliationCashTransaction;
  cash_transactions: ReconciliationCashTransaction[];
  match_type: "single" | "group";
  reconciliation_id?: number;
  matched_at?: string;
}

export interface BankStatementImport {
  id: string;
  original_filename: string;
  content_type?: string;
  file_size: number;
  file_sha256: string;
  row_count: number;
  imported_count: number;
  duplicate_count: number;
  date_from?: string;
  date_to?: string;
  status: "processing" | "processed" | "failed";
  source_type: "manual_pdf" | "manual_spreadsheet" | "manual_image" | "bank_api";
  trust_level: "uploaded_pdf" | "editable_file" | "uploaded_image" | "bank_verified";
  processing_method?: "pdf_text" | "ocr" | "spreadsheet" | "csv";
  parse_message?: string;
  uploaded_by?: number;
  uploaded_by_name?: string;
  created_at: string;
  archived_at?: string;
  can_delete: boolean;
}

export interface BankStatementLinesResponse {
  items: BankStatementLine[];
  total: number;
  counts: {
    suggested: number;
    waiting: number;
    completed: number;
  };
}

// ── API clients ─────────────────────────────────────────────────────────────

export const walletAccountsApi = {
  list: (params?: { owner_type?: string; is_active?: boolean }) =>
    api.get<WalletAccount[]>("/wallet-accounts", { params }).then((r) => r.data),
  get: (id: number) => api.get<WalletAccount>(`/wallet-accounts/${id}`).then((r) => r.data),
  create: (data: unknown) => api.post<WalletAccount>("/wallet-accounts", data).then((r) => r.data),
  update: (id: number, data: unknown) =>
    api.patch<WalletAccount>(`/wallet-accounts/${id}`, data).then((r) => r.data),
  delete: (id: number) => api.delete(`/wallet-accounts/${id}`),
};

export const bankReconciliationApi = {
  accounts: () =>
    api.get<BankReconciliationAccount[]>("/bank-reconciliation/accounts")
      .then((response) => response.data),
  lines: (params: {
    wallet_account_id: number;
    status: "suggested" | "waiting" | "completed";
    start_date?: string;
    end_date?: string;
    search?: string;
    limit?: number;
    offset?: number;
  }) =>
    api.get<BankStatementLinesResponse>("/bank-reconciliation/lines", { params })
      .then((response) => response.data),
  importStatement: (walletAccountId: number, file: File) => {
    const formData = new FormData();
    formData.append("wallet_account_id", String(walletAccountId));
    formData.append("file", file);
    return api.post("/bank-reconciliation/import", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    }).then((response) => response.data);
  },
  imports: (walletAccountId: number, includeArchived = false) =>
    api.get<BankStatementImport[]>("/bank-reconciliation/imports", {
      params: {
        wallet_account_id: walletAccountId,
        include_archived: includeArchived,
      },
    }).then((response) => response.data),
  downloadImport: (importId: string) =>
    api.get<Blob>(`/bank-reconciliation/imports/${importId}/download`, {
      responseType: "blob",
    }).then((response) => response.data),
  archiveImport: (importId: string) =>
    api.post(`/bank-reconciliation/imports/${importId}/archive`)
      .then((response) => response.data),
  restoreImport: (importId: string) =>
    api.post(`/bank-reconciliation/imports/${importId}/restore`)
      .then((response) => response.data),
  deleteImport: (importId: string) =>
    api.delete(`/bank-reconciliation/imports/${importId}`),
  autoMatch: (walletAccountId: number) =>
    api.post("/bank-reconciliation/auto-match", undefined, {
      params: { wallet_account_id: walletAccountId },
    }).then((response) => response.data),
  reconcile: (items: Array<{
    statement_line_id: number;
    cash_transaction_id?: number;
    cash_transaction_ids?: number[];
  }>) =>
    api.post("/bank-reconciliation/reconcile", { items }).then((response) => response.data),
  dismiss: (lineId: number) =>
    api.post(`/bank-reconciliation/lines/${lineId}/dismiss`).then((response) => response.data),
  unreconcile: (lineId: number, reason: string) =>
    api.post(`/bank-reconciliation/lines/${lineId}/unreconcile`, { reason })
      .then((response) => response.data),
};

export const holdersApi = {
  list: (params?: { holder_type?: string; wallet_account_id?: number; is_active?: boolean }) =>
    api.get<Holder[]>("/holders", { params }).then((r) => r.data),
  get: (id: number) => api.get<Holder>(`/holders/${id}`).then((r) => r.data),
  create: (data: unknown) => api.post<Holder>("/holders", data).then((r) => r.data),
  update: (id: number, data: unknown) =>
    api.patch<Holder>(`/holders/${id}`, data).then((r) => r.data),
  delete: (id: number) => api.delete(`/holders/${id}`),
};

export const categoriesApi = {
  list: (type?: string) =>
    api.get<CashflowCategory[]>("/cashflow-categories", { params: { type } }).then((r) => r.data),
  create: (data: unknown) => api.post<CashflowCategory>("/cashflow-categories", data).then((r) => r.data),
  update: (id: number, data: unknown) =>
    api.patch<CashflowCategory>(`/cashflow-categories/${id}`, data).then((r) => r.data),
  delete: (id: number) => api.delete(`/cashflow-categories/${id}`),
};

export const incomeApi = {
  list: (params?: {
    start_date?: string; end_date?: string; status?: string;
    wallet_account_id?: number; holder_id?: number; category_id?: number;
    keyword?: string; limit?: number; offset?: number;
  }) => api.get<IncomeEntry[]>("/income", { params }).then((r) => r.data),
  get: (id: string) => api.get<IncomeEntry>(`/income/${id}`).then((r) => r.data),
  create: (data: unknown) => api.post<IncomeEntry>("/income", data).then((r) => r.data),
  update: (id: string, data: unknown) =>
    api.patch<IncomeEntry>(`/income/${id}`, data).then((r) => r.data),
  delete: (id: string) => api.delete(`/income/${id}`),
};

export const expensesApi = {
  list: (params?: {
    start_date?: string; end_date?: string; status?: string;
    wallet_account_id?: number; holder_id?: number; category_id?: number;
    is_company_expense?: boolean; keyword?: string; limit?: number; offset?: number;
  }) => api.get<ExpenseEntry[]>("/expenses", { params }).then((r) => r.data),
  get: (id: string) => api.get<ExpenseEntry>(`/expenses/${id}`).then((r) => r.data),
  create: (data: unknown) => api.post<ExpenseEntry>("/expenses", data).then((r) => r.data),
  update: (id: string, data: unknown) =>
    api.patch<ExpenseEntry>(`/expenses/${id}`, data).then((r) => r.data),
  delete: (id: string) => api.delete(`/expenses/${id}`),
};

export const payablesApi = {
  list: (params?: { status?: string; due_before?: string; keyword?: string; limit?: number }) =>
    api.get<Payable[]>("/payables", { params }).then((r) => r.data),
  get: (id: string) => api.get<Payable>(`/payables/${id}`).then((r) => r.data),
  create: (data: unknown) => api.post<Payable>("/payables", data).then((r) => r.data),
  update: (id: string, data: unknown) =>
    api.patch<Payable>(`/payables/${id}`, data).then((r) => r.data),
  pay: (id: string, data: { amount: number; account_id?: number; holder_id?: number; paid_date: string }) =>
    api.post<Payable>(`/payables/${id}/pay`, data).then((r) => r.data),
  delete: (id: string) => api.delete(`/payables/${id}`),
};

export const receivablesApi = {
  list: (params?: { status?: string; due_before?: string; keyword?: string; limit?: number }) =>
    api.get<Receivable[]>("/receivables", { params }).then((r) => r.data),
  get: (id: string) => api.get<Receivable>(`/receivables/${id}`).then((r) => r.data),
  create: (data: unknown) => api.post<Receivable>("/receivables", data).then((r) => r.data),
  update: (id: string, data: unknown) =>
    api.patch<Receivable>(`/receivables/${id}`, data).then((r) => r.data),
  receive: (id: string, data: { amount: number; account_id?: number; holder_id?: number; received_date: string }) =>
    api.post<Receivable>(`/receivables/${id}/receive`, data).then((r) => r.data),
  delete: (id: string) => api.delete(`/receivables/${id}`),
};

export const transfersApi = {
  list: (params?: { start_date?: string; end_date?: string; transfer_type?: string; limit?: number }) =>
    api.get<Transfer[]>("/transfers", { params }).then((r) => r.data),
  create: (data: unknown) => api.post<Transfer>("/transfers", data).then((r) => r.data),
  cancel: (id: string) => api.delete(`/transfers/${id}`),
};

export const dashboardApi = {
  getCashflow: (params?: { month?: number; year?: number }) =>
    api.get("/cashflow-dashboard", { params }).then((r) => r.data),
  getSchedule: (params?: { start_date?: string; end_date?: string }) =>
    api.get("/payment-schedule", { params }).then((r) => r.data),
  getLogs: (params?: { resource_type?: string; limit?: number }) =>
    api.get("/activity-logs", { params }).then((r) => r.data),
  getReport: (params: { start_date: string; end_date: string; report_type?: string }) =>
    api.get("/cashflow-report", { params }).then((r) => r.data),
};

export const documentsApi = {
  upload: (formData: FormData) =>
    api.post("/documents/upload", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    }).then((r) => r.data),
  list: (referenceType: string, referenceId: string) =>
    api.get(`/documents/${referenceType}/${referenceId}`).then((r) => r.data),
  delete: (id: string) => api.delete(`/documents/${id}`),
};

File: backend/app/models/cashflow.py
"""SQLAlchemy ORM models for Cash Flow Module"""
from datetime import date, datetime
from decimal import Decimal
from typing import Optional
import enum

from sqlalchemy import (
    BigInteger, Boolean, Date, DateTime, Enum, ForeignKey,
    Integer, Numeric, String, Text, func
)
from sqlalchemy.dialects.postgresql import UUID as PG_UUID, JSONB, INET
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class WalletAccountType(str, enum.Enum):
    bank = "bank"
    cash = "cash"
    ewallet = "ewallet"
    credit_card = "credit_card"
    other = "other"


class MoneyOwnerType(str, enum.Enum):
    company = "company"
    personal = "personal"
    mixed = "mixed"


class HolderType(str, enum.Enum):
    company = "company"
    personal = "personal"
    tax = "tax"
    salary = "salary"
    project = "project"
    reserve = "reserve"
    stock = "stock"
    other = "other"


class CashflowCategoryType(str, enum.Enum):
    income = "income"
    expense = "expense"
    payable = "payable"
    receivable = "receivable"


class EntryStatus(str, enum.Enum):
    pending = "pending"
    completed = "completed"
    cancelled = "cancelled"


class PayableStatus(str, enum.Enum):
    unpaid = "unpaid"
    partial = "partial"
    paid = "paid"
    overdue = "overdue"
    cancelled = "cancelled"


class ReceivableStatus(str, enum.Enum):
    unreceived = "unreceived"
    partial = "partial"
    received = "received"
    overdue = "overdue"
    cancelled = "cancelled"


class TransferType(str, enum.Enum):
    account_to_account = "account_to_account"
    holder_to_holder = "holder_to_holder"
    account_to_holder = "account_to_holder"
    holder_to_account = "holder_to_account"
    owner_withdrawal = "owner_withdrawal"
    owner_advance = "owner_advance"
    salary = "salary"
    dividend = "dividend"


class ApprovalStatus(str, enum.Enum):
    pending = "pending"
    approved = "approved"
    rejected = "rejected"


class CashDirection(str, enum.Enum):
    IN = "in"
    OUT = "out"
    TRANSFER_IN = "transfer_in"
    TRANSFER_OUT = "transfer_out"


def enum_values(enum_class):
    """Persist the enum's public values, not Python member names.

    This matters for CashDirection, whose Python names are upper-case while the
    existing PostgreSQL enum values are lower-case.
    """
    return [member.value for member in enum_class]


# ─── WalletAccount ────────────────────────────────────────────
class WalletAccount(Base):
    __tablename__ = "wallet_accounts"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    account_type: Mapped[WalletAccountType] = mapped_column(
        Enum(WalletAccountType, name="wallet_account_type", create_type=False, values_callable=enum_values), default=WalletAccountType.bank
    )
    owner_type: Mapped[MoneyOwnerType] = mapped_column(
        Enum(MoneyOwnerType, name="money_owner_type", create_type=False, values_callable=enum_values), default=MoneyOwnerType.company
    )
    bank_name: Mapped[Optional[str]] = mapped_column(String(100))
    account_number: Mapped[Optional[str]] = mapped_column(String(50))
    account_holder: Mapped[Optional[str]] = mapped_column(String(200))
    currency: Mapped[str] = mapped_column(String(3), default="THB")
    opening_balance: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    current_balance: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    notes: Mapped[Optional[str]] = mapped_column(Text)
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


# ─── Holder ───────────────────────────────────────────────────
class Holder(Base):
    __tablename__ = "holders"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    holder_type: Mapped[HolderType] = mapped_column(
        Enum(HolderType, name="holder_type", create_type=False, values_callable=enum_values), default=HolderType.company
    )
    owner_type: Mapped[MoneyOwnerType] = mapped_column(
        Enum(MoneyOwnerType, name="money_owner_type", create_type=False, values_callable=enum_values), default=MoneyOwnerType.company
    )
    wallet_account_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("wallet_accounts.id"))
    purpose: Mapped[Optional[str]] = mapped_column(Text)
    opening_balance: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    current_balance: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    responsible_user_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    notes: Mapped[Optional[str]] = mapped_column(Text)
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


# ─── CashflowCategory ─────────────────────────────────────────
class CashflowCategory(Base):
    __tablename__ = "cashflow_categories"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    type: Mapped[CashflowCategoryType] = mapped_column(
        Enum(CashflowCategoryType, name="cashflow_category_type", create_type=False, values_callable=enum_values)
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    parent_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("cashflow_categories.id"))
    color: Mapped[Optional[str]] = mapped_column(String(7))
    icon: Mapped[Optional[str]] = mapped_column(String(50))
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


# ─── IncomeEntry ──────────────────────────────────────────────
class IncomeEntry(Base):
    __tablename__ = "income_entries"

    id: Mapped[str] = mapped_column(PG_UUID(as_uuid=False), primary_key=True, server_default=func.uuid_generate_v4())
    income_date: Mapped[date] = mapped_column(Date, nullable=False)
    document_no: Mapped[Optional[str]] = mapped_column(String(50))
    income_type: Mapped[Optional[str]] = mapped_column(String(50))
    category_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("cashflow_categories.id"))
    customer_name: Mapped[Optional[str]] = mapped_column(String(300))
    description: Mapped[str] = mapped_column(Text, nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    vat_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    withholding_tax: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    net_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    payment_channel: Mapped[Optional[str]] = mapped_column(String(100))
    wallet_account_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("wallet_accounts.id"))
    holder_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("holders.id"))
    status: Mapped[EntryStatus] = mapped_column(
        Enum(EntryStatus, name="entry_status", create_type=False, values_callable=enum_values), default=EntryStatus.pending
    )
    received_date: Mapped[Optional[date]] = mapped_column(Date)
    owner_type: Mapped[MoneyOwnerType] = mapped_column(
        Enum(MoneyOwnerType, name="money_owner_type", create_type=False, values_callable=enum_values), default=MoneyOwnerType.company
    )
    notes: Mapped[Optional[str]] = mapped_column(Text)
    receivable_id: Mapped[Optional[str]] = mapped_column(PG_UUID(as_uuid=False))
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


# ─── ExpenseEntry ─────────────────────────────────────────────
class ExpenseEntry(Base):
    __tablename__ = "expense_entries"

    id: Mapped[str] = mapped_column(PG_UUID(as_uuid=False), primary_key=True, server_default=func.uuid_generate_v4())
    expense_date: Mapped[date] = mapped_column(Date, nullable=False)
    document_no: Mapped[Optional[str]] = mapped_column(String(50))
    expense_type: Mapped[Optional[str]] = mapped_column(String(50))
    category_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("cashflow_categories.id"))
    vendor_name: Mapped[Optional[str]] = mapped_column(String(300))
    description: Mapped[str] = mapped_column(Text, nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    vat_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    withholding_tax: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    net_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    payment_channel: Mapped[Optional[str]] = mapped_column(String(100))
    wallet_account_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("wallet_accounts.id"))
    holder_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("holders.id"))
    is_company_expense: Mapped[bool] = mapped_column(Boolean, default=True)
    status: Mapped[EntryStatus] = mapped_column(
        Enum(EntryStatus, name="entry_status", create_type=False, values_callable=enum_values), default=EntryStatus.pending
    )
    paid_date: Mapped[Optional[date]] = mapped_column(Date)
    owner_type: Mapped[MoneyOwnerType] = mapped_column(
        Enum(MoneyOwnerType, name="money_owner_type", create_type=False, values_callable=enum_values), default=MoneyOwnerType.company
    )
    notes: Mapped[Optional[str]] = mapped_column(Text)
    payable_id: Mapped[Optional[str]] = mapped_column(PG_UUID(as_uuid=False))
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


# ─── Payable ──────────────────────────────────────────────────
class Payable(Base):
    __tablename__ = "payables"

    id: Mapped[str] = mapped_column(PG_UUID(as_uuid=False), primary_key=True, server_default=func.uuid_generate_v4())
    creditor_name: Mapped[str] = mapped_column(String(300), nullable=False)
    creditor_type: Mapped[Optional[str]] = mapped_column(String(50))
    description: Mapped[Optional[str]] = mapped_column(Text)
    issue_date: Mapped[date] = mapped_column(Date, nullable=False)
    due_date: Mapped[Optional[date]] = mapped_column(Date)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    paid_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    expected_account_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("wallet_accounts.id"))
    expected_holder_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("holders.id"))
    category_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("cashflow_categories.id"))
    status: Mapped[PayableStatus] = mapped_column(
        Enum(PayableStatus, name="payable_status", create_type=False, values_callable=enum_values), default=PayableStatus.unpaid
    )
    reference_doc: Mapped[Optional[str]] = mapped_column(String(200))
    notes: Mapped[Optional[str]] = mapped_column(Text)
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    @property
    def remaining_amount(self) -> Decimal:
        return self.total_amount - self.paid_amount


# ─── Receivable ───────────────────────────────────────────────
class Receivable(Base):
    __tablename__ = "receivables"

    id: Mapped[str] = mapped_column(PG_UUID(as_uuid=False), primary_key=True, server_default=func.uuid_generate_v4())
    debtor_name: Mapped[str] = mapped_column(String(300), nullable=False)
    debtor_type: Mapped[Optional[str]] = mapped_column(String(50))
    description: Mapped[Optional[str]] = mapped_column(Text)
    issue_date: Mapped[date] = mapped_column(Date, nullable=False)
    due_date: Mapped[Optional[date]] = mapped_column(Date)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    received_amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    expected_account_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("wallet_accounts.id"))
    expected_holder_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("holders.id"))
    category_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("cashflow_categories.id"))
    status: Mapped[ReceivableStatus] = mapped_column(
        Enum(ReceivableStatus, name="receivable_status", create_type=False, values_callable=enum_values), default=ReceivableStatus.unreceived
    )
    reference_doc: Mapped[Optional[str]] = mapped_column(String(200))
    notes: Mapped[Optional[str]] = mapped_column(Text)
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    @property
    def remaining_amount(self) -> Decimal:
        return self.total_amount - self.received_amount


# ─── Transfer ─────────────────────────────────────────────────
class Transfer(Base):
    __tablename__ = "transfers"

    id: Mapped[str] = mapped_column(PG_UUID(as_uuid=False), primary_key=True, server_default=func.uuid_generate_v4())
    transfer_date: Mapped[date] = mapped_column(Date, nullable=False)
    transfer_type: Mapped[TransferType] = mapped_column(
        Enum(TransferType, name="transfer_type", create_type=False, values_callable=enum_values), default=TransferType.account_to_account
    )
    from_account_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("wallet_accounts.id"))
    from_holder_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("holders.id"))
    to_account_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("wallet_accounts.id"))
    to_holder_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("holders.id"))
    amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    fee: Mapped[Decimal] = mapped_column(Numeric(15, 2), default=0)
    reason: Mapped[Optional[str]] = mapped_column(Text)
    status: Mapped[EntryStatus] = mapped_column(
        Enum(EntryStatus, name="entry_status", create_type=False, values_callable=enum_values), default=EntryStatus.completed
    )
    notes: Mapped[Optional[str]] = mapped_column(Text)
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


# ─── Document ─────────────────────────────────────────────────
class Document(Base):
    __tablename__ = "documents"

    id: Mapped[str] = mapped_column(PG_UUID(as_uuid=False), primary_key=True, server_default=func.uuid_generate_v4())
    reference_type: Mapped[str] = mapped_column(String(50), nullable=False)
    reference_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    file_name: Mapped[str] = mapped_column(String(500), nullable=False)
    file_path: Mapped[str] = mapped_column(String(1000), nullable=False)
    file_type: Mapped[Optional[str]] = mapped_column(String(50))
    file_size: Mapped[Optional[int]] = mapped_column(Integer)
    doc_type: Mapped[Optional[str]] = mapped_column(String(50))
    description: Mapped[Optional[str]] = mapped_column(String(500))
    uploaded_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


# ─── CashTransaction ──────────────────────────────────────────
class CashTransaction(Base):
    __tablename__ = "cash_transactions"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    transaction_date: Mapped[date] = mapped_column(Date, nullable=False)
    direction: Mapped[CashDirection] = mapped_column(
        Enum(CashDirection, name="cash_direction", create_type=False, values_callable=enum_values)
    )
    reference_type: Mapped[str] = mapped_column(String(50), nullable=False)
    reference_id: Mapped[Optional[str]] = mapped_column(PG_UUID(as_uuid=False))
    wallet_account_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("wallet_accounts.id"))
    holder_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("holders.id"))
    amount: Mapped[Decimal] = mapped_column(Numeric(15, 2), nullable=False)
    balance_after: Mapped[Optional[Decimal]] = mapped_column(Numeric(15, 2))
    description: Mapped[Optional[str]] = mapped_column(Text)
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    created_by: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


# ─── ActivityLog ──────────────────────────────────────────────
class ActivityLog(Base):
    __tablename__ = "activity_logs"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    user_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("users.id"))
    action: Mapped[str] = mapped_column(String(50), nullable=False)
    resource_type: Mapped[str] = mapped_column(String(50), nullable=False)
    resource_id: Mapped[Optional[str]] = mapped_column(String(100))
    old_data: Mapped[Optional[dict]] = mapped_column(JSONB)
    new_data: Mapped[Optional[dict]] = mapped_column(JSONB)
    description: Mapped[Optional[str]] = mapped_column(Text)
    company_id: Mapped[int] = mapped_column(Integer, ForeignKey("companies.id"), nullable=False, default=1)
    ip_address = mapped_column(INET)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
